ipu-speed-advantage/benchmarking_gnn_on_ipus_review.ipynb

```diff
@@ -0,0 +1,867 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "ec425747",
+   "metadata": {},
+   "source": [
+    "Copyright (c) 2023 Graphcore Ltd. All rights reserved."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e70cc6e3",
+   "metadata": {},
+   "source": [
+    "# Benchmarking Message Passing on the IPU\n",
+    "\n",
+    "One of the key features of GNNs is message passing where a node's features are passed to its neighbours through some operations. Message passing essentially consists in two operations, namely gathering the node embeddings onto their outbound edges and scattering the result onto the nodes that these edges connect to. There can be a number of additional operations between the gather and scatter, depending on the architecture of the model you are interested in, and therefore it is useful to consider the gather and scatter operations as distinct steps in message passing.\n",
+    "\n",
+    "![Message Passing](./static/gather_scatter.jpg \"Message Passing\")\n",
+    "\n",
+    "In this notebook we will take a look at the performance of the gather and scatter operations on Graphcore hardware in order to understand the performance advantage the IPU offers for these operations. Accelerating these fundamental operations with the IPU also directly translates into performance advantages in message passing and GNNs as a whole.\n",
+    "\n",
+    "Herein, we will:\n",
+    " * Understand how we construct a harness to run benchmarks of particular operations on the IPU,\n",
+    " * Run a sweep of scatter and gather operations on the IPU,\n",
+    " * Take a look at the results of a larger sweep.\n",
+    "\n",
+    "Let's get started!\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4a466dc3",
+   "metadata": {},
+   "source": [
+    "### Running on Paperspace\n",
+    "\n",
+    "The Paperspace environment lets you run this notebook with minimal effort. To improve your experience we preload datasets and pre-install packages, this can take a few minutes. If you experience errors immediately after starting a session, please try restarting the kernel before contacting support. If a problem persists or you want to give us feedback on the content of this notebook, please reach out to us through our community of developers using our [slack channel](https://www.graphcore.ai/join-community) or raise a [GitHub issue](https://github.com/graphcore/examples).\n",
```

Comment:
```suggestion
    "The Paperspace environment lets you run this notebook with minimal effort. To improve your experience we preload datasets and preinstall packages, this can take a few minutes. If you experience errors immediately after starting a session, please try restarting the kernel before contacting support. If a problem persists or you want to give us feedback on the content of this notebook, please reach out to us through our community of developers using our [Slack channel](https://www.graphcore.ai/join-community) or raise a [GitHub issue](https://github.com/graphcore/examples).\n",
```